### PR TITLE
Support parallel kill,rm

### PR DESCRIPTION
**Pull request by WeiZhang555**
Signed-off-by: Zhang Wei zhangwei555@huawei.com

### CODE REVIEW

File: cli/command/container/rm.go [hunk 1]
```diff
@@ -45,13 +45,22 @@ func runRm(dockerCli *command.DockerCli, opts *rmOptions) error {
...
-		if name == "" {
+	options := types.ContainerRemoveOptions{
+		RemoveVolumes: opts.rmVolumes,
+		RemoveLinks:   opts.rmLink,
+		Force:         opts.force,
+	}
+
+	errChan := parallelOperation(ctx, opts.containers, func(ctx context.Context, id string) error {
+		if id == "" {
 			return fmt.Errorf("Container name cannot be empty")
```
Comment (AkihiroSuda) [inline cli/command/container/rm.go]: "Container ID .."?

Comment (WeiZhang555) [inline cli/command/container/rm.go]: Not indeed. It can be either container id or container name, up to input from user. :smile:

**Comment by WeiZhang555** [comment]
ping @thaJeztah @vdemeester @mlaventure

File: cli/command/container/rm.go [hunk 2]
```diff
@@ -45,13 +45,22 @@ func runRm(dockerCli *command.DockerCli, opts *rmOptions) error {
...
-	for _, name := range opts.containers {
-		if name == "" {
+	options := types.ContainerRemoveOptions{
+		RemoveVolumes: opts.rmVolumes,
+		RemoveLinks:   opts.rmLink,
+		Force:         opts.force,
+	}
+
+	errChan := parallelOperation(ctx, opts.containers, func(ctx context.Context, id string) error {
+		if id == "" {
```
Comment (vdemeester) [inline cli/command/container/rm.go]: `nameOrId` :angel: ?

**Review by vdemeester — APPROVED** [review]
One nit comment but overall LGTM 🐸

File: cli/command/container/rm.go (hunk 2 above)
Comment (WeiZhang555) [inline cli/command/container/rm.go]: OK

Comment (WeiZhang555) [inline cli/command/container/rm.go]: Maybe `container` is better? it covers name and id :smile:

Comment (vdemeester) [inline cli/command/container/rm.go]: true :+1:

**Comment by duglin** [comment]
I'm a bit concerned about this PR, but I might be missing something in the code.  How does this impact the UX of the cmds? For example, how do we guarantee that the output isn't overlapping?  Right now with cmds that take multiple containers as args there's a guarantee that the output will always be in a certain order, and it will follow the exact list of container IDs specified on the cmd line. This allows people to easily correlate which bit of output goes with which container. This PR appears to remove that guarantee, meaning the user then has to figure out which output goes with which container.  Do we guarantee that ALL output from each kill/rm has enough uniquely identifying information (like the container name/id) in the output to always allow this correlation?

**Comment by WeiZhang555** [comment]
@duglin don't worry, we have the output order gurantee :smile: . See https://github.com/docker/docker/pull/24761#r73254732 in PR: #24761

I made a generic public function `parallelOperation`, which will execute some function and return result in same order with input.

**Comment by duglin** [comment]
ok cool - wasn't sure from looking at the code. I did notice we show the errors/ID in the right order, but wasn't sure if there would ever be any other output that's sent to stderr/stdout.  Guess not.

**Comment by WeiZhang555** [comment]
@duglin I think the output should be right. What do you think of this? :smile:

**Comment by WeiZhang555** [comment]
What's the current status? ping @mlaventure @thaJeztah

**Comment by mlaventure** [comment]
This LGTM :+1:

**Review by thaJeztah — APPROVED** [review]
Tested this, and works as advertised

LGTM as well

**Comment by thaJeztah** [comment]
Don't think we need documentation changes for this, other than a mention in the changelog

**Pull request merged**